File: _Practice/LeetCode/010. Regular Expression Matching.cpp
// Dynamic Programming
// https://leetcode.com/problems/regular-expression-matching/
// Time: O(n*m), Space: O(n*m)

class Solution {

public:
    bool isMatch(string &s, string &p) { // Top-Down Memoization
        int n = s.size(), m = p.size();
        vector <vector<short>> cache(n + 1, vector<short>(m, -1));

        function<bool(int, int)> dfs = [&](int i, int j) { // nested function (avoiding variable passing)
            if (i <= n && j < m && cache[i][j] >= 0) { // check if it is already in the cache
                return (bool) cache[i][j];
            }
            if (j >= m) { // if we read the pattern string
                return i >= n; // check if we also read the string for matching
            }
            // index i could be technically out of bounds here (that's why cache rows are n+1)
            bool match = (i < n && (s[i] == p[j] || p[j] == '.'));
            // * has the highest precedence (we hendle * first)
            if (j + 1 < m && p[j + 1] == '*') {
                return (bool) (cache[i][j] = (dfs(i, j + 2) // don't use *
                                              || (match && dfs(i + 1, j)))); // use *
            }
            if (match) { // simple match of one character
                return (bool) (cache[i][j] = dfs(i + 1, j + 1));
            }
            return (bool) (cache[i][j] = false);
        };
        return dfs(0, 0);
    }
};

File: 6 – Practice/leet code/010. Regular Expression Matching.java
// https://leetcode.com/problems/regular-expression-matching/
// Dynamic Programming
// Time: O(n*m), Space: O(n*m)

class Solution {
    private String s, p;
    private int n, m;
    private Boolean[][] cache;

    private boolean dfs(int i, int j) {
        if (i <= n && j < m && cache[i][j] != null) { // check if it is already in the cache
            return cache[i][j];
        }
        
        if (j >= m) { // if we read the pattern string
            return i >= n; // check if we also read the string for matching
        }
        // index i could be technically out of bounds here (that's why cache rows are n+1)
        boolean match = i < n && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.');
        // * has the highest precedence (we handle * first)
        if (j + 1 < m && p.charAt(j + 1) == '*') {
            return cache[i][j] = dfs(i, j + 2) // don't use *
                    || (match && dfs(i + 1, j)); // use *
        }
        
        if (match) { // simple match of one character
            return cache[i][j] = dfs(i + 1, j + 1);
        }
        
        return cache[i][j] = false;

    }

    public boolean isMatch(String s, String p) { // Top-Down Memoization
        this.s = s;
        this.p = p;

        this.n = s.length();
        this.m = p.length();

        this.cache = new Boolean[n + 1][m];
        
        return dfs(0, 0);
    }
}